File: utility-benchmarks/python/matrix_mul.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
import utilities as util

config = tf.ConfigProto()
config.gpu_options.allow_growth = True


def matrix_mul(size, warming_up):
    a = tf.random_uniform([size, size], -1, 1)
    b = tf.random_uniform([size, size], -1, 1)
    tf.matmul(a, b)


def run(size, backend):
    if backend == 'gpu' and util.can_use_gpu():
        doWarmup = True
    else:
        assert backend == 'cpu', 'Please specify either cpu or gpu as backend options'
        doWarmup = False

    backend = util.get_backend(backend)
    with tf.device(backend):
        sess = tf.Session()

        if doWarmup:
            sess.run(matrix_mul(size, True))

            #  Add performance benchmarking here!
            sess.run(matrix_mul(size, False))
        else:
            sess.run(matrix_mul(size, False))


if __name__ == "__main__":
    run(10, 'gpu')

File: utility-benchmarks/javascript/matrix_mul.ts
import * as dl from 'deeplearn';

import {BenchmarkTest} from './benchmark';
import * as benchmark_util from './utilities';

export class MatrixMulCPU implements BenchmarkTest {
    async run(size: number): Promise<number> {
        dl.setBackend('cpu');
        const a: dl.Tensor2D = dl.randomUniform([size, size], -1, 1);
        const b: dl.Tensor2D = dl.randomUniform([size, size], -1, 1);
        const start = performance.now();
        dl.matMul(a, b);
        const end = performance.now();
        return end - start;
    }
}

export class MatrixMulGPU implements BenchmarkTest {
    async run(size: number): Promise<number> {
        dl.setBackend('webgl');

        const a: dl.Tensor2D = dl.randomNormal([size, size]);
        const b: dl.Tensor2D = dl.randomNormal([size, size]);

        const benchmark = () => dl.matMul(a, b);

        const time = await benchmark_util.warmupAndBenchmarkGPU(benchmark);

        a.dispose();
        b.dispose();

        return time;
    }
}

export async function run(size, type, params, backend){
    let benchmark: BenchmarkTest = (backend === 'gpu') ? new MatrixMulGPU() : new MatrixMulCPU();
    return await benchmark.run(size)
}
